nntrainer/layers/rnn.cpp

```diff
@@ -54,7 +105,66 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
 }
 
 void RNNLayer::forwarding(bool training) {
-  // NYI
+  Tensor &weight_xh =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getVariableRef();
+  Tensor &bias_h =
+    weightAt(static_cast<int>(RNNParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  Tensor temp;
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor islice = input_.getBatchSlice(b, 1);
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+    float *hidden_next = oslice.getData();
+
+    if (!training) {
+      if (b > 0) {
+        float *hidden_pre =
```

Comment:
I am a little confused. I think the hidden layer is not shared across the batch. Each batch has its own hidden information in a generic RNN.

Reply:
```python
from tensorflow.keras.layers import Input, SimpleRNN
in = Input(shape=(3, 1), batch_size=10)
print(in.shape) # TensorShape([10, 3, 1])
rnn = SimpleRNN(5)(in) # can also enable return_state=True to get the state
print(rnn.shape) # TensorShape([10, 5]) - hidden cell value
```
The hidden state between batch size should not share their values.

Reply:
for the inference, I guess it should. but backwarding, it's not.

Reply:
IMHO, even for inference, it should be not.

Reply:
There are number of points that needs clarification.

I think there are two ways.

1. Keep states for the batch only 🌴 
2. Presist states 🐟 

The one I am referencing suggested to train with 🌴 and inference 🐟, Jijoong told me the implementation is done in 🐟  but later add `stateful=false` as a property.

keras seems to have two options.
https://keras.io/guides/working_with_rnns/#crossbatch-statefulness

Reply:
Stateful and batch are very different concepts.
Stateful talks about keeping the hidden state across different runs.
Batch is about different data points running in parallel.

The behavior of RNN remains the same in inference and train for the batch and stateful. NNTrainer's RNN's default behavior must match other frameworks default behavior to avoid confusion.

> https://keras.io/guides/working_with_rnns/#crossbatch-statefulness

Although the name is `crossbatch`, the state is independent across the batch. This is just a way to selectively reset the state when needed. This does not shares the state across the batch (multiple data points given at once). The `batch` terminology used in the article refers to `iteration`, and it is equivalent to a simple behavior, where we can relinquish the power to reset the hidden state to the user.